packages/react-provider/src/ThemeProvider.tsx

```diff
@@ -0,0 +1,84 @@
+//
+// TODO: should be moved to `react-theme-provider`
+// https://github.com/microsoft/fluentui/pull/16698
+
+import { useMergedRefs } from '@fluentui/react-hooks';
+import { mergeThemes, themeToCSSVariables, PartialTheme, Theme } from '@fluentui/react-theme';
+import { getSlots, makeMergeProps } from '@fluentui/react-utils';
+import * as React from 'react';
+
+import { ThemeContext, useTheme } from './context';
+
+export interface ThemeProviderProps {
+  theme?: PartialTheme | Theme;
+}
+export interface ThemeProviderState {
+  style: React.HTMLAttributes<HTMLElement>['style'];
+  theme: Theme;
+}
+
+const mergeProps = makeMergeProps<ThemeProviderState>();
+
+export function useThemeProviderState(draftState: ThemeProviderState) {
+  const parentTheme = useTheme();
+  const localTheme = draftState.theme;
+
+  draftState.theme = mergeThemes(parentTheme, localTheme);
+  draftState.style = React.useMemo(() => {
+    // TODO: should we consider insertion to head?
+    //       - how to modify, remove styles?
+    //       - SSR rendering
+    return {
```

Comment:
We should inject only variables changed/added by this theme, not the ones consumed via context. It's parent provider's responsibility to inject them.